tests/test_flow/test_reload.py

```diff
@@ -40,216 +40,240 @@ def _write_module(name, contents):
-    tester.expect_values(total=11)
+    flow_tester.expect_values(total=11)
 
     # Major version remains unchanged, cached value is returned
     write_flow_module(10, 2)
-    tester.reload_flow()
-    tester.expect_values(total=11)
+    flow_tester.reload_flow()
+    flow_tester.expect_values(total=11)
 
     # Major version changed, a new result is computed
     write_flow_module(20, 2)
-    tester.reload_flow()
-    tester.expect_values(total=22)
+    flow_tester.reload_flow()
+    flow_tester.expect_values(total=22)
 
 
-def test_assist_versioning_mode(write_module, reload_method):
+def test_assist_versioning_mode(flow_tester):
     def write_flow_module(x: int):
-        write_module(
-            "flow_module",
+        flow_tester.write_tester_flow(
             f"""
-                import bionic as bn
-
-                builder = bn.FlowBuilder('flow')
                 builder.set('core__versioning_mode', 'assist')
 
+                @builder
+                def one():
+                    return 1
+
                 @builder
                 def x():
                     return {x}
+            """,
+        )
+
+    write_flow_module(1)
+    flow_tester.expect_values(one=1, x=1)
+
+    write_flow_module(1)
+    flow_tester.reload_flow()
+    flow_tester.expect_values(one=1, x=1)
+
+    write_flow_module(2)
+    flow_tester.reload_flow()
+    flow_tester.expect_values(one=1)
+    with pytest.raises(CodeVersioningError):
+        flow_tester.get("x")
+
+
+def test_auto_versioning_mode(flow_tester):
+    def write_flow_module(x: int):
+        flow_tester.write_tester_flow(
+            f"""
+                builder.set('core__versioning_mode', 'auto')
 
                 @builder
                 def one():
                     return 1
 
-                flow = builder.build()
+                @builder
+                def x():
+                    return {x}
+
+                @builder
+                def x_doubled(x):
+                    return x * 2
             """,
         )
 
     write_flow_module(1)
-    tester = FlowTester(reload_method)
-    tester.expect_values(one=1, x=1)
-
-    write_flow_module(1)
-    tester.reload_flow()
-    tester.expect_values(one=1, x=1)
+    flow_tester.expect_values(one=1, x=1, x_doubled=2)
 
     write_flow_module(2)
-    tester.reload_flow()
-    tester.expect_values(one=1)
-    with pytest.raises(CodeVersioningError):
-        tester.get("x")
+    flow_tester.reload_flow()
+    flow_tester.expect_values(one=1, x=2, x_doubled=4)
+
+    write_flow_module(3)
+    flow_tester.reload_flow()
+    flow_tester.expect_values(one=1, x=3, x_doubled=6)
 
 
 # Test nondeterministic function marked with the changes_per_run decorator
-def test_changes_per_run_decorator(write_module, reload_method):
+def test_changes_per_run_decorator(flow_tester):
     def write_flow_module():
-        write_module(
-            "flow_module",
+        flow_tester.write_tester_flow(
```

Comment:
WDYT about using `flow.setting('core__persistent_cache__flow_dir', ...)` to modify the flow *after* loading it? This would let you avoid this awkward two-part module writing.

(To elaborate: whenever you set or reload `self.flow` inside `FlowTester`, you also create a modified copy with the appropriate cache setting, and you use that for all your testing.)

Reply:
For `reload` to work, the flow needs to be the first flow built from the builder, and it has not been modified. `flow.setting` breaks this requirement.

Reply:
Right, I was thinking you would store both the original flow and the modified flow. Whenever you need to call `reload`, you call it on the original flow, and then you use `setting` to get the modified reloaded version.

Reply:
Hmm, that would make the test less simple, and it would also change the nature of the flow as well; we are testing a flow that exist in the `test_reload.py`, not in the temporary `flow_module.py`.

Reply:
Personally I think it would make the tests simpler, because each test would be writing a complete module that defines its own `builder`, instead of having to know that `write_tester_flow` magically adds it to the code. And I don't think it changes the nature of the flow in a meaningful way; the key thing is whether you see the old or new behavior, and using `setting` to configure the cache won't change that.

That said, it's up to you!